src/types/set.rs

```diff
@@ -71,96 +71,248 @@ impl PySet {
+        K: ToPyObject;
+
+    /// Removes and returns an arbitrary element from the set.
+    fn pop(&self) -> Option<Bound<'py, PyAny>>;
+
+    /// Returns an iterator of values in this set.
+    ///
+    /// # Panics
+    ///
+    /// If PyO3 detects that the set is mutated during iteration, it will panic.
+    fn iter(&self) -> BoundSetIterator<'py>;
+}
+
+impl<'py> PySetMethods<'py> for Bound<'py, PySet> {
+    #[inline]
+    fn clear(&self) {
+        unsafe {
+            ffi::PySet_Clear(self.as_ptr());
+        }
+    }
+
+    #[inline]
+    fn len(&self) -> usize {
+        unsafe { ffi::PySet_Size(self.as_ptr()) as usize }
+    }
+
+    fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    fn contains<K>(&self, key: K) -> PyResult<bool>
+    where
+        K: ToPyObject,
+    {
+        fn inner(set: &Bound<'_, PySet>, key: Bound<'_, PyAny>) -> PyResult<bool> {
             match unsafe { ffi::PySet_Contains(set.as_ptr(), key.as_ptr()) } {
                 1 => Ok(true),
                 0 => Ok(false),
                 _ => Err(PyErr::fetch(set.py())),
             }
         }
 
-        inner(self, key.to_object(self.py()))
+        let py = self.py();
+        inner(self, key.to_object(py).into_bound(py))
     }
 
-    /// Removes the element from the set if it is present.
-    ///
-    /// Returns `true` if the element was present in the set.
-    pub fn discard<K>(&self, key: K) -> PyResult<bool>
+    fn discard<K>(&self, key: K) -> PyResult<bool>
     where
         K: ToPyObject,
     {
-        fn inner(set: &PySet, key: PyObject) -> PyResult<bool> {
+        fn inner(set: &Bound<'_, PySet>, key: Bound<'_, PyAny>) -> PyResult<bool> {
             match unsafe { ffi::PySet_Discard(set.as_ptr(), key.as_ptr()) } {
                 1 => Ok(true),
                 0 => Ok(false),
                 _ => Err(PyErr::fetch(set.py())),
             }
         }
 
-        inner(self, key.to_object(self.py()))
+        let py = self.py();
+        inner(self, key.to_object(py).into_bound(py))
     }
 
-    /// Adds an element to the set.
-    pub fn add<K>(&self, key: K) -> PyResult<()>
+    fn add<K>(&self, key: K) -> PyResult<()>
     where
         K: ToPyObject,
     {
-        fn inner(set: &PySet, key: PyObject) -> PyResult<()> {
+        fn inner(set: &Bound<'_, PySet>, key: Bound<'_, PyAny>) -> PyResult<()> {
             err::error_on_minusone(set.py(), unsafe {
                 ffi::PySet_Add(set.as_ptr(), key.as_ptr())
             })
         }
 
-        inner(self, key.to_object(self.py()))
+        let py = self.py();
+        inner(self, key.to_object(py).into_bound(py))
     }
 
-    /// Removes and returns an arbitrary element from the set.
-    pub fn pop(&self) -> Option<PyObject> {
-        let element =
-            unsafe { PyObject::from_owned_ptr_or_err(self.py(), ffi::PySet_Pop(self.as_ptr())) };
+    fn pop(&self) -> Option<Bound<'py, PyAny>> {
+        // FIXME should this return PyResult?
```

Comment:
From the description in https://docs.python.org/3/c-api/set.html#c.PySet_Pop, I'd say that the current type is right. We could check that the error is indeed a `KeyError` and panic for `SystemError`, i.e. when this was somehow not called on a set, but I think we don't panic due to this elsewhere either.

So while technically we could make this return `PyResult<Option<Bound<PyAny>>>`, this would only catch the invalid situation when `PySet` does reference something that is not actually set and I think this is of limited usefulness.

Reply:
Yes, agreed 👍. Thanks for thinking about this; I'd forgotten I'd left that FIXME in the code!